### PR TITLE
Adding typescript types

### PR DESCRIPTION
**Pull request by calumleask**
Not all types defined yet but left TODOs where they are missing.
- Updated eslint and fixed errors regarding 'hasOwnProperty' usage.
- Returning 'this' from setters in the Prop class for consistency
- Removing the 'width' option from Wrld.Polyline since it is overridden by the 'weight' option which also has a default. Wrld.Polyline is also undocumented. There are examples for it but they don't try to use the width option.

### CODE REVIEW

File: types/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,434 @@
...
+        tiltDegrees: number;
+        animate: boolean;
+        durationSeconds: number;
+        allowInterruption: boolean;
+    };
+}
+
+// eslint-disable-next-line no-redeclare
+interface Map extends L.Map {
+    getAltitudeAtLatLng(latLng: L.LatLngExpression): number;
```
Comment (iainbethune) [inline types/index.d.ts:67]: Is this correct.  I think getAltitudeAtLatLng can only accept L.LatLng or { lat: number; lng: number } but L.LatLngExpression also includes [number, number]

File: types/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,434 @@
...
+    elevationMode: ElevationMode;
+    indoorMapId: IndoorMapId;
+    indoorMapFloorId: IndoorMapFloorId;
+};
+
+class Marker extends L.Marker {
+    constructor(latlng: L.LatLngExpression, options?: MarkerOptions);
+    getElevation(): number;
+    setElevation(elevation : number): this;
+    getElevationMode(): this;
```
Comment (iainbethune) [inline types/index.d.ts:119]: Returns ElevationMode?

File: types/index.d.ts (hunk 1 above)
Comment (iainbethune) [inline types/index.d.ts:67]: Is this correct? I think getAltitudeAtLatLng can only accept L.LatLng or { lat: number; lng: number } but L.LatLngExpression also includes [number, number]

File: types/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,434 @@
...
+type PropOptions = {
+    elevation?: number;
+    elevationMode?: ElevationMode;
+    indoorMapId?: IndoorMapId;
+    indoorMapFloorId?: IndoorMapFloorId;
+    headingDegrees?: number;
+};
+
+class Prop {
+    constructor(name: string, geometryId: string, location: L.LatLngExpression, options?: PropOptions);
```
Comment (iainbethune) [inline types/index.d.ts:199]: Similar to getAltitudeAtLatLng() - our code assumes we can address a lat and lng property of the object so we can't take a LatLngTuple

File: types/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,434 @@
...
+    addTo(map: Map): this;
+    remove(): this;
+    getLocation(): L.LatLng;
+    setLocation(latLng: L.LatLngExpression): this;
```
Comment (iainbethune) [inline types/index.d.ts:203]: Same

File: types/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,434 @@
...
+    type PointData = [number, number] | [number, number, number] | {
+        latLng?: L.LatLngExpression;
+        weight?: number;
+    } | {
+        [dataCoordProperty: string]: L.LatLngExpression;
+        [dataWeightProperty: string]: number;
+    };
+
+    // eslint-disable-next-line no-unused-vars
+    type DesnsityStop = {
```
Comment (iainbethune) [inline types/index.d.ts:231]: Typo: DensityStop

File: types/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,434 @@
...
+namespace indoorMapFloorOutlines {}
+
+/* Wrld */
+
+function map(element: HTMLElement | string, apiKey: string, options?: MapOptions): Map;
+function marker(latLng: L.LatLngExpression, options?: MarkerOptions): Marker;
+function popup(options?: PopupOptions, source?: L.Layer): Popup;
+function polygon(latlngs: L.LatLngTuple[] | L.LatLngTuple[][], options?: PolygonOptions): Polygon;
+function polyline(latlngs: L.LatLngExpression[], options?: PolylineOptions): Polyline;
+function prop(name: string, geometryId: string, location: L.LatLngExpression, options?: PropOptions): Prop;
```
Comment (iainbethune) [inline types/index.d.ts:404]: Same comment re: LatLngExpression here

**Review by iainbethune — CHANGES_REQUESTED** [review]
Good stuff!  I hadn't realised you could also define string enums as types, which is a nice bonus :)

Only substantive issue is where we don't accept a LatLngExpression.  Think this can easily be solved by creating a L.LatLng from the parameter and then we can happily refer to .lng and .lat without issue.

File: types/index.d.ts (hunk 2 above)
Comment (calumleask) [reply types/index.d.ts]: Fixed now

File: types/index.d.ts (hunk 1 above)
Comment (calumleask) [reply types/index.d.ts:67]: Fixed by converting latLng in SpacesApi by wrapping latLng(latLng)

File: types/index.d.ts (hunk 3 above)
Comment (calumleask) [reply types/index.d.ts:199]: This is already wrapped by L.latLng https://github.com/wrld3d/wrld.js/blob/types/src/public/prop.js#L8

File: types/index.d.ts (hunk 4 above)
Comment (calumleask) [reply types/index.d.ts:203]: This is already wrapped by L.latLng https://github.com/wrld3d/wrld.js/blob/types/src/public/prop.js#L24

File: types/index.d.ts (hunk 5 above)
Comment (calumleask) [reply types/index.d.ts]: Fixed now

File: types/index.d.ts (hunk 6 above)
Comment (calumleask) [reply types/index.d.ts:404]: Shouldn't be an issue as mentioned above

**Review by iainbethune — APPROVED** [review]

**Pull request merged**